GEP-0068: Gateway API Extension for Gardener Shoot Clusters (#69)
gardener/enhancements: geps/0068-gateway-api-extension/README.md

# GEP-68: Gateway API Extension for Gardener Shoot Clusters

## Summary

The Kubernetes ecosystem is converging on the [Gateway API](https://gateway-api.sigs.k8s.io/)

> as the long-term successor to the `Ingress` resource. [Gateway API graduated to GA

Reviewer: `successor` gives me the impression the `Ingress` support will be removed, which I understand is not the case. I would rephrase the statement to something like 

```suggestion
as the next generation implementation of the `Ingress` resource. [Gateway API graduated to GA
```
or 

```suggestion
as the evolution of the `Ingress` resource. [Gateway API graduated to GA
```

but will leave it up to you.

with v1.0 in October 2023](https://kubernetes.io/blog/2023/10/31/gateway-api-ga/) and has since received broad implementation support
across the CNCF landscape. With [GEP-57](../0057-replace-nginx-ingress-shoot-addon-with-traefik-extension/README.md)
already establishing a Traefik-based replacement for the retired Ingress NGINX
shoot addon, Gardener users now need a dedicated, first-class option for

> Gateway API workloads in their shoot clusters.

Reviewer: What about the Gardener users in the garden and seed domains? Don't they also need support for Gateway API? What would be the effort to add support for `garden` and `seed` class to the extension?

Author: Currently we only want to focus on the shoot side. Are you aware of some components/users who needs it?

Reviewer: As long as it is not supported in the garden and seeds, nobody will use it. But any application like plutono, vali, prometheus, etc. can be exposed via Gateway API, or any other extension deploying workload in the seed that needs to be exposed. 

To be fair, I am also not aware of anybody asking for Gateway API in the shoot, but here we are and I do not see a reason to not enable it for seeds and garden.

Reviewer 2: Currently, we use istio for exposure of resources in `garden` and `seed`. This is sufficient for now.

Adding support for `garden` and `seed` scenarios in the extension should be possible, but is not part of the initial scope as far as I understand it. Do you think that it is mandatory for the minimal viable scope?

Reviewer: I am mostly asking the extension to be designed in a way it can easily be extended with support for garden and seed classes in the future, not to be necessary included in the initial implementation.

> After a detailed evaluation of several unknown open-source Gateway API

Reviewer: `detailed evaluation of several unknown`, is this a typo? If not, it is contradicting. When something is evaluated in details it should be well known what it is, not unknown. Considering you are listing names of implementations on the next line, I believe it is some kind of typo.

implementations — Envoy Gateway, Traefik, Istio, Kgateway, Cilium, Kong, and
NGINX Gateway Fabric — informed by the upstream conformance benchmark
[gateway-api-bench](https://github.com/howardjohn/gateway-api-bench), this GEP
recommends **[Envoy Gateway](https://gateway.envoyproxy.io/)** as the
implementation shipped to Gardener shoot clusters, with **Traefik Gateway API**

> documented as the runner-up.

Reviewer: Does this mean the extension will support multiple gateway implementations? As a Gardener user I indeed may benefit from different implementations and use different features when needed, but this clears the path to add support for yet another one. What would be the requirements the extension to support another implementation?

Reviewer: Please, add link to the Traefik Gateway API documentation/source repo.

Author: This extension will only support envoy gateway api implementation but there could be more extensions like one for istio etc...

The extension follows the standard Gardener extension contract (controller
registration, `ManagedResource`-based deployment, admission webhooks) and is
deliberately scoped narrowly: it manages (installs and updates) the Gateway
API CRDs in the shoot, deploys the Envoy Gateway control plane and the Envoy

> data-plane proxies, and installs a `GatewayClass` named `envoy-gateway` that

Reviewer: Where are the proxies deployed, in the control plane or in the data plane?

Author: The envoy proxies are the data plane, and they run inside the shoot cluster not in the seed control plane.

shoot owners can reference from their `Gateway` and `*Route` objects. (Envoy

> Gateway's helm chart does not ship a `GatewayClass`; the extension creates

Reviewer: Do you plan the implementation to be based on (the official) helm chart? Usage of helm charts should be avoided, instead the resources should be defined as go types. ref: https://github.com/gardener/gardener/blob/master/docs/development/component-checklist.md

Author: the shoot resources will be implemented as go types, I've updated it in the GEP to make it clear.

  now implement Gateway API. Gardener users who do not have a supported
  Gateway API option in their shoots either roll their own (creating a
  fragmented, unsupported landscape) or stay on `Ingress` and accept its
  limitations.

* **Choice and forward compatibility**: `Ingress` remains a fully supported
  API and is not going away. However, some cluster owners will want to
  adopt Gateway API for its richer feature set. Providing a first-class
  extension gives shoot owners the freedom to choose — they can stay on
  `Ingress`, move to Gateway API, or run both side by side. For those who
  do decide to migrate, the transition can happen incrementally and
  per-shoot rather than as a big-bang break.

* **Optional role separation**: Some Gardener shoots are shared by multiple
  teams where platform admins and application developers have distinct
  responsibilities. Gateway API's persona-based resource model
  (`GatewayClass` → `Gateway` → `*Route`) enables a separation of concerns
  that `Ingress` does not offer natively. For single-team shoots or simpler
  setups where this split is unnecessary, `Ingress` remains a
  straightforward and perfectly valid choice.

> * **Vendor neutrality and conformance**: Gateway API has a published

Reviewer: Does the extension plan to run the conformance tests regularly to ensure the conformance of the provided solution?

Author: There are no plans yet.

  conformance suite, and workloads that stick to the standard-channel
  resources (`HTTPRoute`, `GRPCRoute`, etc.) are portable across conformant
  implementations in principle. In practice, however, most non-trivial
  deployments will rely on implementation-specific extensions — Envoy
  Gateway's `EnvoyPatchPolicy`/`EnvoyExtensionPolicy`, Istio's
  `EnvoyFilter`, etc. — which limits real-world portability. The
  conformance suite nonetheless provides a useful baseline: the standard
  routing semantics are well-defined, and simple-to-moderate use cases
  transfer across implementations without modification.

### Goals

1. Introduce `gardener-extension-shoot-envoy-gateway` as an extension in the
   [Gardener GitHub organisation](https://github.com/gardener).
2. Ship **Envoy Gateway** as the bundled Gateway API implementation. The

>    extension installs a `GatewayClass` named `envoy-gateway` bound to the

Reviewer: How the shoot owner can influence the gateway class configuration? For example, I may want the workload exposed via the class to be published in a private network only.

Author: The GatewayClass itself is cluster-scoped and owned by the extension a shoot owner cannot and shouldn't edit it. Instead, a shoot owner controls exposure per Gateway, using standard Gateway API / Envoy Gateway mechanisms. For your "private network only" case, the relevant lever is the load-balancer annotation on the Gateway's generated Service.

4. Integrate with Gardener's standard resource-management, observability,
   and lifecycle mechanisms (`ManagedResource`, heartbeat, metrics, VPA/HPA).
5. Provide an opinionated baseline configuration that works out of the box
   on every cloud provider Gardener supports, while allowing users to
   override behaviour through the extension's `providerConfig`.
6. Enable safe, incremental rollout: restrict the extension initially to
   shoots with `purpose: evaluation`, mirroring the rollout strategy of
   GEP-57.
7. Document a clear migration path from `Ingress` (both legacy NGINX and
   the GEP-57 Traefik extension) to `HTTPRoute`, building on the upstream
   [Migrating from Ingress](https://gateway-api.sigs.k8s.io/guides/getting-started/migrating-from-ingress/)
   guide.

### Non-Goals

1. This GEP does **not** propose migrating Gardener core
   (`gardener/gardener`) to Gateway API. Gardener core is currently in the
   process of replacing its built-in NGINX-based ingress with [another
   solution](https://github.com/gardener/gardener/issues/13448); an internal switch from `Ingress` to Gateway
   API is not on the roadmap and is not what this GEP is about. This GEP

>    strictly concerns user-facing ingress *inside* the shoot cluster.

Reviewer: While it is true Gardener itself is not on the road to adopt the Gateway API, it has made substantial progress on the extensibility of the garden and seed clusters, having support for Gateway API may be a prerequisite for other extensions to expose their workloads from the garden/seed clusters. 

I think it makes sense to not put a hard restriction this extension to be limited only to shoots, but keep it open to support Gardens and Seeds (respectively it should not be named `shoot-envoy-gateway` but rather `envoy-gateway`).

Author: Is there already a use case for that?

Reviewer: Take a look at all of the extensions implementing support for Garden/Seed class - they had not the use case, but now they have it. It is better to design the extension in a way it can easily implement support for more extension classes in the future instead of considering only the shoot class.

2. This GEP does **not** deprecate or remove the GEP-57
   `gardener-extension-shoot-traefik` extension. Both extensions are
   designed to coexist; an operator can offer either, both, or neither.
3. This GEP does **not** prescribe a service mesh. Gateway API has a separate

>    GAMMA initiative for east-west mesh routing; this is out of scope here.

Reviewer: Can you motivate this decision? It seems to be part of the standard API since v1.1.0, ref: https://gateway-api.sigs.k8s.io/docs/introduction/#gateway-api-for-service-mesh-the-gamma-initiative

Author: I've updated the GEP and motivated this decision.

4. This GEP does **not** introduce multiple competing `GatewayClass`
   objects. The extension installs a single `GatewayClass` named
   `envoy-gateway` bound to the Envoy Gateway controller. Users may install
   additional `GatewayClass` objects pointing at other implementations
   independently of this extension.
5. This GEP does **not** cover network policy, mTLS automation, or
   advanced traffic policies (rate limiting, JWT auth, WAF). These can be
   layered on top via the chosen implementation's policy CRDs but are not
   exercised by the extension itself in the initial release.

## Proposal

Introduce `gardener-extension-shoot-envoy-gateway` as a new extension in the
Gardener GitHub organisation. The extension follows the well-established
[Gardener Extension Concept](https://gardener.cloud/docs/gardener/extensions/)
and implements the `Extension` reconciler contract.

When enabled on a Shoot, the extension:

1. Reconciles `ManagedResource` objects in the shoot's control-plane namespace
   on the seed. The `gardener-resource-manager` then applies the contained

>    manifests into the shoot cluster (CRDs, RBAC, the Envoy Gateway control-plane

Reviewer: Might be misunderstanding on my side, but is it possible to run the Envoy Gateway control-plane in the seed?

Author: Running the Envoy Gateway control plane in the seed is technically conceivable (Gardener already hosts shoot control-plane components there), but it breaks down on xDS connectivity: the shoot's data-plane Envoy proxies are xDS clients that must dial back to the control plane's xDS server, which would require a shoot→seed connection that Gardener's networking model (reversed VPN, seed→shoot only) deliberately does not provide. On top of that, Envoy Gateway injects the in-cluster service DNS (envoy-gateway.<ns>.svc.cluster.local:18000) into each proxy's bootstrap and assumes a single API server for both watching resources and provisioning proxies, so a split topology would need cross-cluster addressing and an unsupported upstream patch. Keeping the control plane co-located with the data plane in the shoot makes that xDS link shoot-local and avoids exposing it across the seed/shoot boundary.

   Deployment, the `envoy-gateway` `GatewayClass`, etc.).
2. Registers an admission webhook that validates `Shoot` objects enabling the
   extension to enforce the evaluation-purpose scope constraint and to
   validate the `EnvoyGatewayConfig` provider config.
3. Exposes a `/metrics` endpoint on the extension controller for Gardener's

>    monitoring stack to scrape.

Reviewer: Will you integrate the envoy own observability signals with the Gardener observability stack?

Author: Currently not planned.

4. Participates in the Gardener heartbeat protocol to report extension health.

The extension type identifier is **`shoot-envoy-gateway`** (referenced in
`spec.extensions[].type` of the Shoot manifest).

#### Deployment Topology: Seed vs Shoot

To avoid ambiguity about *what runs where*, the table below lists every
component the extension is responsible for and the cluster it ends up in.

| Component | Cluster | Notes |
|-----------|---------|-------|

> | `gardener-extension-shoot-envoy-gateway` controller | Seed (per seed) | Watches `Extension` objects of type `shoot-envoy-gateway` and reconciles `ManagedResource` objects. This is the Gardener-side "operator". |

Reviewer: ```suggestion
| `gardener-extension-shoot-envoy-gateway` controller | Seed (per seed) | Watches `Extension` objects of type `shoot-envoy-gateway` and deploys `ManagedResource` objects. This is the Gardener-side "operator". |
```

GRM is the controller manager that actually reconciles MR.

| Admission webhook | Garden runtime / virtual garden | Validates `Shoot` resources at admission time. Standard Gardener admission deployment pattern. |
| Gateway API CRDs | Shoot | Standard channel `gateway.networking.k8s.io/v1`; optionally experimental channel. Delivered via `ManagedResource`. |
| Envoy Gateway CRDs (`EnvoyProxy`, `BackendTrafficPolicy`, `ClientTrafficPolicy`, `SecurityPolicy`, …) | Shoot | Required for the Envoy Gateway control plane to function. |

> | Envoy Gateway **control plane** (Deployment, Service, RBAC, PDB, optional VPA/HPA) | Shoot | Runs as Pods inside the shoot. Translates `Gateway`/`*Route` resources into Envoy xDS configuration. |

Reviewer 3: In particular use-cases I would like to have the possibility to use different implementation of GWAPI, but still keep the APIs managed by Gardener. Would that be possible without EnvoyGateway installation? It's mostly under Kyma scope, where we manage our own Istio inside shoot cluster through operator. I understand that installing istio alongisde its own GatewayClass is a possibility, but keeping relatively low footprint and is also very important.

Author: Thanks for raising this -  I thought about it, but I'd like to keep the extension scoped to "Envoy Gateway + Gateway API CRDs" as one self-contained bundle rather than expose  a CRDs-only mode.

 The reasoning mirrors the discussion further up this PR around `manageCRDs` / `experimentalFeatures`: if Gardener ships a Gateway API extension, it's cleanest that Gardener also owns the corresponding CRDs *and* the implementation behind the `envoy-gateway` `GatewayClass`. Once we let users bring their own implementation while the extension still reconciles the CRDs, we're back to two owners for the same CRD set (this extension vs. the Kyma Istio operator, both of which want to install/update the `gateway.networking.k8s.io` CRDs) - exactly the ownership conflict we want to avoid.

  For the Kyma use case, the intended path stays as it is on any other Kubernetes distribution today: don't enable `shoot-envoy-gateway`, and let the in-shoot Istio operator install both the Gateway API CRDs and its `istio` `GatewayClass` - that's a scenario the Istio operator already handles. The footprint of the CRDs themselves is minimal; the real footprint (Envoy Gateway control plane + per-`Gateway` proxy pods + LB Services) only appears when this extension is actually enabled, and if Kyma already has Istio doing the job, there's no reason to enable it.

  If down the line there's broad demand for a "Gateway-API-CRDs-only" extension decoupled from any implementation, I'd rather do that as a separate, dedicated extension (its own GEP, its own lifecycle) than dilute the scope of this one. WDYT?

Reviewer 3: Thanks for the explanation. We also noticed another problem with the CRD-only reconciliation. We can have problems with GWAPI version handling when istio does not support the version that extension manages.

Having the operator manage CRDs and implementation is the cleanest approach.

| `GatewayClass` (`envoy-gateway`) | Shoot | Created by this extension via `ManagedResource`. Bound to controller `gateway.envoyproxy.io/gatewayclass-controller`. Envoy Gateway's helm chart does not ship a `GatewayClass`, so the extension provides one. |

> | Envoy **data plane** (proxy Pods) | Shoot | Spawned by the Envoy Gateway control plane in response to user-created `Gateway` objects. Each `Gateway` gets its own Envoy Deployment + Service inside the shoot. |

Reviewer: `control plane` vs `control-plane`, I see both are being used, I am fine with either of them, but please stick to only one.

> | LoadBalancer `Service` per `Gateway` | Shoot | Provisioned by the cloud-provider load-balancer controller running inside the shoot, exactly like an `Ingress`-mode LB today. |

Reviewer: What is running in the shoot? If it is the cloud-provider load-balancer controller, that is probably wrong assumption as they are running in the seed. Actually, how is the LoadBalancer service provisioned, is it done implicitly by the envoy controllers when they reconcile the Gateway resource or it should be explicitly provisioned externally to the envoy controllers, e.g. by the extension?

Author: In the shoot runs: Envoy Gateway control plane + data plane, CRDs, GatewayClass, and the per-Gateway LB Services. 
LB provisioning: implicit - Envoy Gateway creates the LoadBalancer Service on Gateway reconcile; the extension does not provision it externally. The extension only lays down the control plane, CRDs, GatewayClass, and NetworkPolicies that permit external traffic to reach the proxies. CCM runs in control plane.

   process, which violates the Gateway API spec's namespace isolation
   expectations and creates a noisy-neighbour risk in multi-tenant shoots.
2. Its status reporting is slow (~180 seconds for status updates on large
   route sets) and it fails to apply large route volumes — both
   deal-breakers for shoots with many application teams.

### Notes/Constraints/Caveats

* **Gateway API CRDs are installed cluster-wide.** The extension installs the
  Gateway API standard-channel CRDs (`gateway.networking.k8s.io/v1`:
  `GatewayClass`, `Gateway`, `HTTPRoute`, `GRPCRoute`, `ReferenceGrant`).
  These CRDs are non-namespaced and will exist in any shoot where the
  extension is enabled, even if no `Gateway` resource is created.

* **Experimental channel is opt-in.** The `gateway.networking.k8s.io/v1alpha2`
  experimental CRDs (`TCPRoute`, `TLSRoute`, `UDPRoute`, `BackendTLSPolicy`)
  are not installed by default. They can be enabled by setting
  `spec.experimentalFeatures: true` in the extension's provider config
  (`spec.extensions[].providerConfig` on the Shoot, see [API](#api)).
  Operators should be aware that experimental APIs may change in

>   backwards-incompatible ways.

Reviewer: Do you plan to be more vocal about this risk? For example, an admission controller can return warnings on shoot create/update operations?

Author: Yes that is true but if you enable experimentalFeatures you should generally be aware that the feature could be dropped entirely and that the API is not stable. Do you want an explicit warning for that?

* **Envoy Gateway CRDs (e.g. `EnvoyProxy`, `BackendTrafficPolicy`,
  `ClientTrafficPolicy`, `SecurityPolicy`) are also installed.** These are
  namespaced and required for advanced Envoy-specific configuration. Users

>   who only consume the standard Gateway API surface can ignore them.

Reviewer: Is it possible these CRDs to be in conflict with other envoy based solution installed in the cluster that brings own version of the CRDs spec?

Author: For `gateway.envoyproxy.io` only if another envoy gateway API installation in the shoot exists.

* **`GatewayClass` is `envoy-gateway`.** The extension installs a single
  `GatewayClass` named `envoy-gateway` bound to controller
  `gateway.envoyproxy.io/gatewayclass-controller`. Envoy Gateway's helm
  chart does not create a `GatewayClass` on its own, so the extension is
  responsible for it. Users reference it from their `Gateway` objects via
  `spec.gatewayClassName: envoy-gateway`.

* **Coexistence with shoot-traefik.** Both the GEP-57 Traefik ingress
  extension and this extension can be installed in the same shoot. They
  do not conflict at the IngressClass / GatewayClass level. See
  [Coexistence with `shoot-traefik`](#coexistence-with-shoot-traefik).

* **Scope is currently restricted to `purpose: evaluation` shoots.** This
  constraint is enforced by the admission webhook and exists to allow the
  extension to mature in low-risk environments before being opened up to
  development and production shoots.

### Risks and Mitigations

| Risk | Likelihood | Impact | Mitigation |
|------|-----------|--------|------------|

> | Envoy Gateway memory leak under route/config churn (reported in [gateway-api-bench](https://github.com/howardjohn/gateway-api-bench)) | Medium | Medium | The control-plane pod is configured with a memory limit and a VPA `Recreate` updatePolicy so leaked memory is reclaimed by periodic restart. The issue is actively tracked upstream; the extension pins to a release where mitigations have landed. |

Reviewer: Gardener pod autoscaling best practices advise not to use memory limits, see https://github.com/gardener/gardener/blob/master/docs/usage/autoscaling/shoot_pod_autoscaling_best_practices.md#summary. 

Software solution are never expected to be bug-free, I mean the GEP could have been proposed at a time when it is not known that a memory leak exists and then such implementation details would not be included. I do not see a reason this fixed in time bug to be part of the GEP, sooner or later it will be fixed.

> | Initial-request errors during bootstrap (also reported upstream) | Medium | Low | Readiness gates on the data-plane Envoy pods; documentation advises users to use a startup-probe-based health check from their LB. |

Reviewer: What will exactly the extension do to apply the mitigation? Will it keep track of additional Gateways created by the shoot owners reconciled by the same envoy gateway implementation?

| Gateway API CRD conflicts if user pre-installed them | Low | High | The extension is opt-in per shoot, so users enabling it consent to Gardener managing the Gateway API CRDs. CRDs are part of the shoot `ManagedResource` and applied via server-side apply; existing CRDs are updated idempotently without field-ownership conflicts. |
| Limited annotation/feature parity with NGINX or Traefik Ingress | High | Medium | Documented migration guide: most NGINX `Ingress` annotations have a direct `HTTPRoute` filter or Envoy `BackendTrafficPolicy` equivalent. Users requiring features not yet expressible in standard Gateway API are advised to remain on the Traefik extension until the experimental channel covers their case. |
| Operators end up with three concurrent ingress paths in one shoot (legacy NGINX, Traefik, Gateway API) | Medium | Medium | Documentation strongly recommends a single ingress path per shoot in production. The evaluation-purpose scope of both extensions limits the blast radius during the rollout phase. |

> | Ecosystem churn: Gateway API evolves rapidly (v1.5 as of early 2026, v1.6 in progress) | High | Low | The extension declares conformance against a specific Gateway API release in its release notes and bumps deliberately, not automatically. |

Reviewer: Is it planned to support multiple version so that shoot owners have the possibility to firstly qualify the new version for their workload and just after that upgrade their production clusters?

Author: Yes, the versioning for extensions will be in another GEP.

          tag: latest
        replicaCount: 1
        resources:
          requests:
            cpu: 50m
            memory: 192Mi
        vpa:
          enabled: true
          resourcePolicy:
            minAllowed:
              memory: 128Mi
          updatePolicy:
            updateMode: Recreate
  resources:
  - clusterCompatibility:
    - shoot
    kind: Extension
    lifecycle:
      delete: BeforeKubeAPIServer
      migrate: AfterKubeAPIServer
      reconcile: AfterKubeAPIServer
    type: shoot-envoy-gateway
    workerlessSupported: false
```

or as `ControllerDeployment` and `ControllerRegistration`:

```yaml
# ControllerDeployment
apiVersion: core.gardener.cloud/v1beta1
kind: ControllerDeployment
metadata:
  name: gardener-extension-shoot-envoy-gateway
helm:
  rawChart: <base64-encoded Helm chart>
```

```yaml
# ControllerRegistration
apiVersion: core.gardener.cloud/v1beta1
kind: ControllerRegistration
metadata:
  name: shoot-envoy-gateway
spec:
  resources:
    - kind: Extension
      type: shoot-envoy-gateway
      globallyEnabled: false   # opt-in per shoot
      lifecycle:

>         reconcile: AfterKubeAPIServer
>         delete:   BeforeKubeAPIServer

Reviewer: ```suggestion
        reconcile: AfterKubeAPIServer
        delete:   BeforeKubeAPIServer
        migrate: AfterKubeAPIServer
```

for completeness

  deployment:
    deploymentRefs:
      - name: gardener-extension-shoot-envoy-gateway
```

The extension controller is deployed per seed and watches `Extension` objects
of type `shoot-envoy-gateway`.

### API

> The extension introduces a new API group `envoy-gateway.extensions.gardener.cloud`

Reviewer: ```suggestion
The extension introduces a new API group `envoy.gateway.extensions.gardener.cloud`
```

This would allow to logically group other gateway implementations under the `gateway.extensions.gardener.cloud` domain, similarly to how it is done for provider extensions.

Author: That's true. But are there any plans to support more than one solution? There is written that we might reconsider Kgateway at a later point in time but it will then be more like a replacement. If you think that there will come more extensions supporting a different gateway API implementation, I will change it. What's your opinion on that?

Reviewer: Gardener initially supported only CoreOS as node operating system, today it supports Garden Linux, SuSE CHost, Ubuntu, Flatcar, Debian, and potentially many more that are not known to the Gardener community. Same is for the networking, initially it was calico only, today cilium is also available. 

My idea is to be pragmatic and open for future development in the area, do not put restrictions that could prevent graceful evolution of the topic.

> with a single versioned kind `EnvoyGatewayConfig`:

Reviewer: Considering the API Group is envoy gateway specific, the kind could be named simply `Config` to avoid duplications.

Author: The kind is frequently shown without its group. Admission-webhook error messages, controller logs, kubectl api-resources, event messages, and RBAC rules often surface the kind on its own. `EnvoyGatewayConfig` is self-describing there; a bare Config is meaningless out of context. Thus I would still prefer this name. We also have a similar naming pattern in other extensions. For example registry.extensions.gardener.cloud -> RegistryConfig, service.cert.extensions.gardener.cloud -> CertConfig ...

Reviewer: Ok, fair with me. Thanks for the reasoning!

```yaml
apiVersion: envoy-gateway.extensions.gardener.cloud/v1alpha1
kind: EnvoyGatewayConfig
spec:
  # Log level for both control-plane and data-plane: debug | info | warn | error (default: info)

>   logLevel: info

Reviewer: Will those logs be scraped and persisted in the logging stack provisioned by Gardener? Too high verbosity can have negative influence on the network traffic and storage footprint, so it should be evaluated whether `debug` level is acceptable.

Author: A debug log level is useful to have, it is not the default and should only be used when required. Not to have it feels wrong to me.

Reviewer: Ok, landscape operators can disallow it via out-of-band validating webhook if it is causing issues for them.

  # `gateway.envoyproxy.io/v1alpha1.EnvoyProxy` reference.
  envoyProxyDefaults:
    accessLogging: true
```

Resource sizing for both the Envoy Gateway control-plane and the Envoy
data-plane is handled by Gardener-managed HPA and VPA in line with how
other shoot-system components are sized; the API intentionally does not
expose replica counts or resource requests. The Gateway API CRDs are
always managed by the extension — opting into the extension means
delegating CRD lifecycle to Gardener.

This object is embedded as `providerConfig` in the Shoot's
`spec.extensions[].providerConfig` field. Internal type conversion and
defaulting are handled by the API machinery registered under
`envoy-gateway.extensions.gardener.cloud/v1alpha1`.

A Shoot enabling the extension looks like this:

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: my-shoot
  namespace: garden-my-project
spec:
  purpose: evaluation
  extensions:
  - type: shoot-envoy-gateway
    providerConfig:
      apiVersion: envoy-gateway.extensions.gardener.cloud/v1alpha1
      kind: EnvoyGatewayConfig
      spec:
        logLevel: info

>         experimentalFeatures: false

Reviewer 3: If I understand correctly, configuration of the Gateway API controller would be configured during creation or edit of the Shoot cluster. Given that users would like to cover more operator-like approach for the reconciliation of extension components, would it be possible to follow the similar pattern to DNSProvider or Certificates? I'm talking about introducing shoot-specific CR that lets me enable GWAPI reconciliation.

Author: The DNSProvider/Certificate pattern you're referring to works differently: those extensions install a controller in the shoot that watches user-created CRs inside the shoot itself, so any tenant with RBAC on those CRs can drive reconciliation without touching the Shoot object. Gateway API already gives us exactly that shape natively - Gateway, HTTPRoute, GRPCRoute, ReferenceGrant etc. are the shoot-scoped CRs, and Envoy Gateway reconciles them without any Shoot edit. So the "enable GWAPI reconciliation from inside the shoot" flow is already covered by creating Gateway objects. What lives in providerConfig is only the extension-installation-level settings (log level, experimental channel opt-in, etc.) - i.e. things that a shoot operator, not a tenant, decides. That intentionally mirrors how shoot-traefik, shoot-cert-service, shoot-dns-service etc. shape their providerConfig, and keeps a single source of truth for the extension's lifecycle in the Shoot.
Introducing an additional in-shoot CR (say EnvoyGateway) on top of that would either duplicate the same settings or split them across two objects, so I'd prefer to keep the current model.

Reviewer 3: The thing is, we mostly configure connectivity in Kyma on shoot level, because we deploy operator into a cluster which then installs istio. The idea was to manage the extension behavior from inside the cluster, so we could build around the GWAPI lifecycle. If that's not the case we have to find another solution.

Reviewer 4: I don't think customizing the extension's configuration from within the shoot cluster fits into the general API model. This only works via the `Shoot` resource in the garden cluster. You could, however, introduce connectivity from workload in your shoot cluster to the garden cluster (and manipulate the `Shoot` resource from there), but there is no native approach.

  # ... remaining shoot spec
```

To enable the experimental Gateway API CRDs, set
`spec.experimentalFeatures: true` in the provider config above.

### Provider Modes

The extension does not expose multiple back-end providers. Unlike the
`shoot-traefik` extension, where `KubernetesIngress` vs `KubernetesIngressNGINX`
modes serve a clear migration purpose, Gateway API is a single, well-defined
spec — there is no analogous compatibility shim to model. The implementation
behind the `envoy-gateway` `GatewayClass` is always Envoy Gateway in this release.

A user wishing to run a different implementation (for example Istio for mesh
integration) can do so independently of this extension by installing their
own `GatewayClass`. Such side-by-side installations are out of scope for the
extension's own validation but are not blocked.

### Admission Webhook

> A `ValidatingWebhookConfiguration` is registered at path

Reviewer: Looks like this about the shoor resource validation, please clarify that the webhook is registered in the garden cluster.

`/webhooks/validate-shoot-envoy-gateway`. It validates:

> 1. **Purpose restriction**: The extension may only be enabled (`disabled:

Reviewer: This should be a temporary validation until the extension graduates, so if you want it to be part of the GEP, it should be explained that this restriction will be eventually lifted in the future.

   false`) on Shoots whose `spec.purpose` is `evaluation`. Adding the
   extension to a non-evaluation shoot is rejected with a descriptive error.
2. **Provider config schema**: If a non-nil `providerConfig` is supplied, it
   must decode successfully as an `EnvoyGatewayConfig` object. Unknown fields
   cause a validation error (strict decoding).
3. **Field constraints**: `spec.logLevel` must be one of the accepted enum
   values.

### Lifecycle Management

The extension interacts with Gardener's lifecycle protocol as follows:

| Phase | Behaviour |
|-------|-----------|
| Reconcile | Creates/updates a `ManagedResource` in the shoot namespace on the seed with all shoot-cluster resources (Gateway API CRDs, Envoy Gateway CRDs, control-plane Deployment, Service, RBAC, PDB, optional HPA/VPA, and the `envoy-gateway` `GatewayClass`). Waits for the `ManagedResource` to become healthy before marking the `Extension` as reconciled. |

> | Delete (extension disabled, shoot kept) | The `Delete` reconciler reads the `Cluster` resource and checks `cluster.Shoot.DeletionTimestamp`. When it is **nil** (shoot stays), the extension refuses to remove itself while user-owned `Gateway` objects still exist in the shoot (to prevent silent traffic loss). The admission webhook surfaces this as a validation error on the `Shoot` update. Once the user has cleaned up their `Gateway`/`*Route` objects, the `ManagedResource` is deleted and the extension waits up to 5 minutes for managed objects to disappear. |

Reviewer: How is `user-owned Gateway object` identified? The extension is managing one Gateway and one GatewayClass, are those resources also expected to be deleted by the user in advance?

> The admission webhook surfaces this as a validation error on the `Shoot` update.

Can you please elaborate on this? Which admission webhook? How are operations scoped to the shoot executed in the shoot or in the seed surfacing via an admission webhook in the Shoot resource? 

> the `ManagedResource` is deleted and the extension waits up to 5 minutes for managed objects to disappear.

How is the extension about to implement this behavior? If the controller gets interrupted after the ManagedResource is gone, how will it know it has to wait some time for manage objects to disappear? Also, what are these managed objects that are not part of the ManagedResource or would allow the ManagedResource deletion to complete before they are gone?

Author: Updated the file

| Delete (shoot deletion) | When `cluster.Shoot.DeletionTimestamp` is **non-nil**, the entire shoot is going away, so the "Gateways still exist" guard is bypassed — blocking would only leak the shoot. The extension's `Extension` resource is reconciled with `lifecycle.delete: BeforeKubeAPIServer`, so the `ManagedResource` (Envoy Gateway control plane, CRDs, the `envoy-gateway` `GatewayClass`, EnvoyProxy/HTTPRoute/Gateway instances) is torn down before the shoot's API server is removed. The cloud-provider `LoadBalancer` Services that were created for each `Gateway` are deleted as part of the shoot's normal `Service` cleanup, freeing the underlying load balancers. No manual cleanup of `Gateway` objects is required from the user. |
| Heartbeat | Extension controller participates in the Gardener heartbeat protocol and reports liveness. |

> | Metrics | Prometheus metrics are exposed on port 8080 under `/metrics`; Gardener's monitoring stack can scrape them via `ServiceMonitor`. The Envoy data-plane and Envoy Gateway control-plane also expose Prometheus metrics that are scraped via separate `ServiceMonitor` objects. |

Reviewer: If the extension is not running anything in the seed for the shoot, what metrics are about to be scraped on port 8080?

Author: Updated the GEP

| VPA/HPA | VPA and HPA manifests are provided for both the Envoy Gateway control-plane and the Envoy data-plane pods in the shoot cluster (VPA enabled by default; HPA enabled for the data-plane). Sizing is fully Gardener-managed — there is no user-facing knob for replica count or resource requests. The extension controller on the seed has its own VPA configuration (see [Extension Registration](#extension-registration)). |

### Coexistence with `shoot-traefik`

A shoot may have both `shoot-traefik` and `shoot-envoy-gateway` enabled. The
two extensions reconcile disjoint resources:

| Resource | Owner |
|----------|-------|
| `IngressClass: nginx` / `IngressClass: traefik` | `shoot-traefik` |
| `GatewayClass: envoy-gateway` | `shoot-envoy-gateway` (registered by Envoy Gateway itself) |
| Traefik Deployment, IngressRoute CRDs | `shoot-traefik` |
| Envoy Gateway Deployment, Gateway API CRDs, EnvoyProxy CRDs | `shoot-envoy-gateway` |

There is no IP/port conflict at the `Service` level because each extension
provisions its own `LoadBalancer` Service. Nonetheless, running both in
parallel doubles the LB cost and the cognitive load; documentation will

> recommend choosing one path per shoot in production.

Reviewer: Can you add more details why only one should be used? Kubernetes does not impose a limitation to have only one service of type LoadBalancer in the cluster, and this is not known to cause cognitive load. Having two distinct solutions for network routing should not be a problem as they should not share component on the network path. 

I can only see the backend/workload being exposed via multiple paths as some kind of duplication, but this is completely in the responsibility of the workload owner, Gardener and Gardener extensions should provide the tools, but leave to the shoot/workload owner how to use them based on their need/liking.

Author: Updated GEP

| C3 | **Architectural correctness**: control- vs data-plane separation, per-namespace Gateway isolation | Multi-team shoots require strong tenant isolation |
| C4 | **Status reporting timeliness** | Slow status updates break GitOps workflows |
| C5 | **Scale**: number of Gateways, Routes, backends supported on a single control plane | A shoot can host hundreds of routes |
| C6 | **Project governance**: vendor neutrality, CNCF status, multi-vendor maintainers | Avoid vendor lock-in |
| C7 | **Operational footprint**: control-plane components, dependency on a service mesh, CRD count | Operability inside a Gardener shoot |
| C8 | **NGINX/Ingress migration story**: support for users coming from `Ingress` | Backwards compatibility with GEP-57 Traefik ingress users |
| C9 | **L4 routing** (`TCPRoute`, `TLSRoute`, `UDPRoute`) | Gardener users running databases, gRPC, custom protocols |
| C10 | **Already in use within Gardener** | Reuse operational knowledge |

### Candidates Considered

The candidate list includes several common implementations that ship a usable
Gateway API standard-channel controller as of late 2025. Implementations

> shipping only the Ingress API are excluded.

Reviewer: What is meant by ingress API here?

Author: The Ingress API means the classic Kubernetes Ingress resource (`networking.k8s.io/v1` Ingress && IngressClass)

Reviewer: Ok, and here we are in GEP about the Gateway API, I am a bit confused when the ingress API is being mentioned so often in this proposal.

  Per-namespace Gateway isolation respected.
* **Conformance**: Full standard channel, partial experimental channel.
  No correctness issues reported in the upstream benchmark beyond a memory
  leak under churn.
* **Governance**: CNCF, hosted under the Envoy organisation. Multi-vendor
  maintainers (Tetrate, VMware, AIS, Bloomberg, others).
* **L4**: `TCPRoute`, `TLSRoute`, `UDPRoute` all supported.
* **Dependencies**: None beyond Kubernetes and the Gateway API CRDs.

#### 2. [Traefik Gateway API](https://doc.traefik.io/traefik/routing/providers/kubernetes-gateway/)

Runner-up. Rejected primarily on architectural correctness in multi-tenant
scenarios, on slow status reconciliation, and on benchmarked failures with
large route volumes.

* **Throughput** (gateway-api-bench): ~217k qps — solid mid-pack.
* **Architecture concern**: The upstream benchmark notes that Traefik
  *"consolidates all Gateways across namespaces unsafely"*, meaning a
  single Traefik process serves Gateways from multiple namespaces. This
  is essentially how most `Ingress` controllers operate and may be
  perfectly acceptable for single-team shoots, but it conflicts with

>   Gateway API's explicit namespace-isolation model and creates a

Reviewer: Can you add reference to this namespace isolation model?

Author: Updated the GEP

  on its own once installed; no manual `GatewayClass` provisioning is needed.
  (Envoy Gateway behaves the same way with `envoy-gateway`.)
* **Why rejected**: Istio bundles a full service mesh (sidecars, mTLS, mesh
  control plane). Installing it merely to get a Gateway API ingress imposes
  a significant operational footprint on every shoot. Users who *want* a
  mesh in their shoot can install Istio independently; its `istio`
  `GatewayClass` will be registered automatically and they can use it from
  their `Gateway` resources. This extension targets the lighter, mesh-free
  use case.

#### 4. [Kgateway](https://kgateway.dev/) (formerly Gloo Gateway, k8sgateway)

Best raw throughput in the benchmark but ruled out on maturity and
governance.

* **Throughput** (gateway-api-bench): ~400k qps — the highest of the field.
* **Architecture**: Clean separation, Envoy-based data plane, no
  correctness issues reported.
* **Why rejected**: The project rebranded from Gloo Gateway to Kgateway in
  2024 and entered the CNCF as a sandbox project shortly after. The CNCF
  sandbox stage is too early for a default Gardener shipping decision.

>   Re-evaluation is appropriate once the project graduates to incubation.

Reviewer: Is it suggested to extend this extension or create new one with support for Kgateway?

Author: Each gateway API implementation should be in it's own extension. If a and b would be fulfilled we would build a new one for Kgateway.

   Operational knowledge built on Envoy Gateway transfers to those
   neighbouring products. Traefik knowledge does not.

The remaining argument in Traefik's favour — that GEP-57 already runs
Traefik in shoots — does not carry over to the Gateway API surface. The
GEP-57 extension exists to provide a smooth migration off the retired
NGINX Ingress controller, and its value lies in NGINX annotation
translation, which is wholly an Ingress-API feature. Gateway API users do
not benefit from that translation. Therefore the operational-reuse argument
applies only to the binary, not to the user-facing API — and even there,
the two extensions can coexist (see
[Coexistence with `shoot-traefik`](#coexistence-with-shoot-traefik)).

For these reasons the extension ships Envoy Gateway. Operators who prefer
Traefik for the Gateway API surface can install it independently in their
shoots and register a separate `GatewayClass`; this GEP does not block
them.

## Future Enhancements

* **Promote out of `purpose: evaluation`.** Once the extension has soaked in
  evaluation shoots and the upstream Envoy Gateway memory-leak issue is
  resolved, the admission-webhook restriction will be relaxed to allow

>   `development` and `production` shoots.

Reviewer: ```suggestion
  resolved, the admission-webhook restriction will be removed.
```

There is also `testing` purpose, I doubt these shoots needs to be restricted to not use the extension.

* **Feature gates** for opt-in experimental Gateway API features
  (`TCPRoute`, `BackendTLSPolicy`, mesh GAMMA bindings) without requiring a
  new extension release.

> * **Implementation-version decoupling.** In the initial release, the Envoy

Reviewer: Is this the version of the envoy binaries or the version of the Gateway API?

Author: It's the envoy gateway implementation version/release.

  Gateway version is pinned in the extension binary. A future iteration
  will allow operators to manage a catalog of Envoy Gateway versions with

>   lifecycle classifications (aligned with

Reviewer: How do you plan to approach the fact that particular Gateway API (envoy) version may not be compatible with the kubernetes version of the cluster? One example is the `ReferenceGrant` API being moved to k8s, this eventually will introduce dependency between the envoy gateway, gateway API and kubernetes versions.

Author: Will be addressed by the Versioning GEP

    - [Why Envoy Gateway Over Traefik Gateway API](#why-envoy-gateway-over-traefik-gateway-api)
  - [Future Enhancements](#future-enhancements)
  - [Drawbacks](#drawbacks)
  - [Alternatives](#alternatives)
    - [1. Do Nothing — Rely on `shoot-traefik` Plus User-Installed Gateway API](#1-do-nothing--rely-on-shoot-traefik-plus-user-installed-gateway-api)
    - [2. Pick Traefik Gateway API for Operational Reuse](#2-pick-traefik-gateway-api-for-operational-reuse)
    - [3. Pick Istio for Best Conformance](#3-pick-istio-for-best-conformance)
    - [4. Pick Kgateway for Best Performance](#4-pick-kgateway-for-best-performance)
    - [5. Cilium Gateway API for CNI Co-location](#5-cilium-gateway-api-for-cni-co-location)
    - [6. NGINX Gateway Fabric for "Familiar" NGINX Branding](#6-nginx-gateway-fabric-for-familiar-nginx-branding)
    - [7. Kong Gateway](#7-kong-gateway)
    - [8. Use Gateway API CRDs Without a Bundled Implementation](#8-use-gateway-api-crds-without-a-bundled-implementation)

## Summary

The Kubernetes ecosystem is converging on the [Gateway API](https://gateway-api.sigs.k8s.io/)
as the long-term successor to the `Ingress` resource. Gateway API graduated to GA

> with v1.0 in October 2023 and has since received broad implementation support

Reviewer 2: ```suggestion
as the long-term successor to the `Ingress` resource. [Gateway API graduated to GA
with v1.0 in October 2023](https://kubernetes.io/blog/2023/10/31/gateway-api-ga/) and has since received broad implementation support
```

across the CNCF landscape. With [GEP-57](../0057-replace-nginx-ingress-shoot-addon-with-traefik-extension/README.md)
already establishing a Traefik-based replacement for the retired Ingress NGINX
shoot addon, Gardener users now need a dedicated, first-class option for
Gateway API workloads in their shoot clusters.

> After a detailed evaluation of all major open-source Gateway API

Reviewer 2: `all major` is a strong statement. Maybe `several known` would also be sufficient.
WDYT?

implementations — Envoy Gateway, Traefik, Istio, Kgateway, Cilium, Kong, and
NGINX Gateway Fabric — informed by the upstream conformance benchmark
[gateway-api-bench](https://github.com/howardjohn/gateway-api-bench), this GEP
recommends **[Envoy Gateway](https://gateway.envoyproxy.io/)** as the
implementation shipped to Gardener shoot clusters, with **Traefik Gateway API**
documented as the runner-up.

The extension follows the standard Gardener extension contract (controller
registration, `ManagedResource`-based deployment, admission webhooks) and is
deliberately scoped narrowly: it manages (installs and updates) the Gateway
API CRDs in the shoot, deploys the Envoy Gateway control plane and the Envoy
data-plane proxies, and installs a `GatewayClass` named `envoy-gateway` that
shoot owners can reference from their `Gateway` and `*Route` objects. (Envoy
Gateway's helm chart does not ship a `GatewayClass`; the extension creates
one, using the controller name
`gateway.envoyproxy.io/gatewayclass-controller`.)

## Motivation

> Gardener's current ingress story is anchored on the legacy `Ingress` resource

Reviewer 2: `legacy` seem a bit hard. Maybe `mature`/`proven` may be better words that have less negative connotation.

> and — going forward — on the Traefik-based replacement defined in GEP-57.

Reviewer 2: Let's add a link to GEP-57.

Author: I've added one link in the first occurrence of `GEP-57` do you want that I add it to each occurrence?

Reviewer 2: It is not necessary to have it linked everywhere, but using the links at least once helps navigating between related proposals.

> and — going forward — on the Traefik-based replacement defined in GEP-57.

Reviewer 2: Actually, the traefik extension is not a replacement for `Ingress` resources, but rather an implementation of an `Ingress` controller. Let's adapt the sentence.

Both options serve the same purpose: a single resource type for HTTP host/path
routing. The shortcomings of `Ingress` are well known:

* The Kubernetes `Ingress` API has been frozen since 2020. New routing features
  (header rewrites, traffic splitting, mirroring, multi-protocol routing,
  cross-namespace references) cannot be expressed without controller-specific
  annotations, which destroy portability.
* No native L4 (TCP/TLS-passthrough) or gRPC routing.
* No clear separation of concerns between platform admins (who provision

>   load balancers) and application teams (who attach routes).

Reviewer 2: It may make sense to add some references/links if we assume that this is well known/documented. For example, https://gateway-api.sigs.k8s.io/guides/getting-started/migrating-from-ingress/ might be a starting point that includes at least some of the topics.

* `GatewayClass` — declared by the infrastructure provider (analogous to
  `StorageClass`).
* `Gateway` — provisioned by the platform/cluster admin; describes the L4/L7
  listener (port, protocol, TLS, allowed namespaces).
* `HTTPRoute`, `GRPCRoute`, `TLSRoute`, `TCPRoute`, `UDPRoute` — owned by
  application teams; attach to a `Gateway` and describe routing rules.
* `ReferenceGrant` — explicit cross-namespace authorisation for backend
  references.

Key problems this GEP addresses:

* **Ecosystem convergence**: Most major service-mesh and ingress vendors
  (Istio, Cilium, Envoy Gateway, Traefik, Kong, NGINX, Kgateway, HAProxy)
  now implement Gateway API. Gardener users who do not have a supported
  Gateway API option in their shoots either roll their own (creating a
  fragmented, unsupported landscape) or stay on `Ingress` and accept its
  limitations.

* **Forward-compatible migration**: Workloads built today on the new Traefik
  ingress extension or on legacy NGINX `Ingress` resources will eventually
  need a path to Gateway API. Providing a first-class extension now means
  this migration can happen incrementally and per-shoot, instead of as a

>   big-bang break.

Reviewer 2: This paints a picture of `Ingress` going away, which is not the case. Maybe we should rephrase it. Having an extension supporting gateway API certainly provides more options to cluster owners to determine on their own whether they want to stay on `Ingress` or use gateway API.

* **Multi-tenancy and role separation**: Gardener shoots are commonly used
  by multiple teams. Gateway API's persona-based resource split aligns much
  better with how Gardener users organise platform vs. application

>   ownership inside a shoot than annotation-driven `Ingress` does.

Reviewer 2: It is certainly possible that a shoot cluster is used by multiple teams, but it is also possible that it is only used by one team. Gateway API allows to split some responsibilities, which is not possible with `Ingress`. However, this split is not alway required. In those cases `Ingress` still might be better suited.

Overall, this bullet point seems overly one-sided and should be revised from my point of view.

* **Vendor neutrality and conformance**: Gateway API has a published
  conformance suite. Selecting a conformant implementation gives Gardener
  users portable routing semantics — workloads written against `HTTPRoute`
  in a Gardener shoot will behave the same way in any other conformant

>   cluster.

Reviewer 2: Not sure if reality really lives up to this expectation. Istio provides `EnvoyFilter` and envoy gateway has `EnvoyPatchPolicy`/`EnvoyExtensionPolicy`, which allow specifying behaviour beyond the gateway API standard. It is yet to be seen whether a direct migration will really work for anything, but a hello-world scenario.

* **L4 + L7 unified**: Several Gardener users today bolt MetalLB or
  cloud-provider LBs in front of `Ingress` controllers to handle TCP/UDP.
  Gateway API's `TCPRoute`/`TLSRoute`/`UDPRoute` collapse this into a single

>   programming model.

Reviewer 2: Not sure if MetalLB is really used that widely, probably cloud-provider LBs would be sufficient.

Furthermore, I think that the unified model is only useful if an L7 loadbalancer/load balancer controller from the corresponding infrastructure is in use, but then we would not need the envoy gateway extension. Hence, is this bullet point relevant for the proposal?

   extension installs a `GatewayClass` named `envoy-gateway` bound to the
   Envoy Gateway controller (`gateway.envoyproxy.io/gatewayclass-controller`);
   Envoy Gateway's helm chart does not ship a `GatewayClass` itself, so the
   extension is responsible for creating one. Shoot users reference it from
   their `Gateway` resources via `spec.gatewayClassName: envoy-gateway`.
3. Install (or reconcile) the Gateway API standard channel CRDs
   (`gateway.networking.k8s.io/v1`) in the shoot cluster.
4. Integrate with Gardener's standard resource-management, observability,
   and lifecycle mechanisms (`ManagedResource`, heartbeat, metrics, VPA/HPA).
5. Provide an opinionated baseline configuration that works out of the box
   on every cloud provider Gardener supports, while allowing users to
   override behaviour through the extension's `providerConfig`.
6. Enable safe, incremental rollout: restrict the extension initially to
   shoots with `purpose: evaluation`, mirroring the rollout strategy of
   GEP-57.
7. Document a clear migration path from `Ingress` (both legacy NGINX and
   the GEP-57 Traefik extension) to `HTTPRoute`.

### Non-Goals

> 1. This GEP does **not** propose migrating Gardener core
>    (`gardener/gardener`) to Gateway API. Gardener core is currently in the
>    process of replacing its built-in NGINX-based ingress with another
>    `Ingress`-based solution; an internal switch from `Ingress` to Gateway
>    API is not on the roadmap and is not what this GEP is about. This GEP
>    strictly concerns user-facing ingress *inside* the shoot cluster.

Reviewer 2: ```suggestion
1. This GEP does **not** propose migrating Gardener core
   (`gardener/gardener`) to Gateway API. Gardener core is currently in the
   process of replacing its built-in NGINX-based ingress with [another
   solution](https://github.com/gardener/gardener/issues/13448); an internal switch from `Ingress` to Gateway
   API is not on the roadmap and is not what this GEP is about. This GEP
   strictly concerns user-facing ingress *inside* the shoot cluster.
```

behalf of the data path**. The seed only hosts the Gardener extension
controller. All ingress traffic, all xDS reconciliation, and all CRD storage
is shoot-local.

### Selected Implementation: Envoy Gateway

Following the evaluation in [Evaluation of Gateway API Implementations](#evaluation-of-gateway-api-implementations),
the extension ships **Envoy Gateway** as the implementation behind the
`envoy-gateway` `GatewayClass`. Envoy Gateway was selected over Traefik
Gateway API and the other candidates because of:

* **Performance**: ~325k qps in the upstream
  [gateway-api-bench](https://github.com/howardjohn/gateway-api-bench) at 512
  connections, ~1.5x Traefik's throughput on the same hardware.
* **Architectural cleanliness**: Proper control-plane / data-plane separation,
  per-namespace Gateway isolation that respects the Gateway API spec.
* **Conformance**: Full support for the Gateway API standard channel; no
  reported correctness issues in the upstream benchmark beyond a known
  memory leak under churn (see [Risks and Mitigations](#risks-and-mitigations)).
* **Foundation**: Built on Envoy, the same data plane used by Istio,

>   Kgateway, and most large-scale production gateway deployments — meaning

Reviewer 2: Let' add links to envoy, istio and kgateway.

>   the data path itself is the most battle-tested codebase in the field.

Reviewer 2: I would argue that `most battle-tested` would fit better to nginx as it predates envoy by more than 10 years and is used extensively in the internet.

> * **CNCF stewardship**: Hosted under the Envoy organisation in CNCF, with

Reviewer 2: Let's add a link.

  multi-vendor maintainers — no single-vendor lock-in.

Traefik Gateway API was a strong runner-up but rejected for two concrete
reasons documented in the upstream benchmark: (1) it consolidates Gateways
across namespaces into a shared data-plane process, which violates the
Gateway API spec's namespace isolation expectations and creates a noisy-
neighbour risk in multi-tenant shoots; and (2) its status reporting is slow
(~180 seconds for status updates on large route sets) and it fails to apply
large route volumes — both deal-breakers for shoots with many application

> teams.

Reviewer 2: Please format the section so that the two arguments a better readable, i.e. properly aligned.

### Notes/Constraints/Caveats

* **Gateway API CRDs are installed cluster-wide.** The extension installs the
  Gateway API standard-channel CRDs (`gateway.networking.k8s.io/v1`:
  `GatewayClass`, `Gateway`, `HTTPRoute`, `GRPCRoute`, `ReferenceGrant`).
  These CRDs are non-namespaced and will exist in any shoot where the
  extension is enabled, even if no `Gateway` resource is created.

* **Experimental channel is opt-in.** The `gateway.networking.k8s.io/v1alpha2`
  experimental CRDs (`TCPRoute`, `TLSRoute`, `UDPRoute`, `BackendTLSPolicy`)
  are not installed by default. They can be enabled via
  `spec.experimentalFeatures: true` in `EnvoyGatewayConfig`. Operators should

>   be aware that experimental APIs may change in backwards-incompatible ways.

Reviewer 2: I am confused. I thought that the gateway API CRDs are deployed by the extension. Therefore, I would have expected the option to enable the experimental gateway API CRDs in its provider configuration, but not in a custom resource from envoy gateway itself.
Please clarify.

| Gateway API CRD conflicts if user pre-installed them | Low | High | CRDs are part of the shoot `ManagedResource` and applied via server-side apply; existing CRDs are updated idempotently without field-ownership conflicts. The extension's CRD reconciliation is opt-out via `spec.manageCRDs: false` for users who manage CRDs themselves. |
| Limited annotation/feature parity with NGINX or Traefik Ingress | High | Medium | Documented migration guide: most NGINX `Ingress` annotations have a direct `HTTPRoute` filter or Envoy `BackendTrafficPolicy` equivalent. Users requiring features not yet expressible in standard Gateway API are advised to remain on the Traefik extension until the experimental channel covers their case. |
| Operators end up with three concurrent ingress paths in one shoot (legacy NGINX, Traefik, Gateway API) | Medium | Medium | Documentation strongly recommends a single ingress path per shoot in production. The evaluation-purpose scope of both extensions limits the blast radius during the rollout phase. |

> | Ecosystem churn: Gateway API adds new GA features in v1.2 / v1.3 | High | Low | The extension declares conformance against a specific Gateway API release in its release notes and bumps deliberately, not automatically. |

Reviewer 2: Maybe this is an easter egg, but is the current Gateway API release not `v1.5`? Why are we referencing `v1.2` and `v1.3` here? The community is fast-moving.

   cause a validation error (strict decoding).
3. **Field constraints**: `spec.controlPlaneReplicas` and
   `spec.dataPlaneReplicas` must be ≥ 1; `spec.logLevel` must be one of the
   accepted enum values.

### Lifecycle Management

The extension interacts with Gardener's lifecycle protocol as follows:

| Phase | Behaviour |
|-------|-----------|
| Reconcile | Creates/updates a `ManagedResource` in the shoot namespace on the seed with all shoot-cluster resources (Gateway API CRDs, Envoy Gateway CRDs, control-plane Deployment, Service, RBAC, PDB, optional HPA/VPA, and the `envoy-gateway` `GatewayClass`). Waits for the `ManagedResource` to become healthy before marking the `Extension` as reconciled. |
| Delete (extension disabled, shoot kept) | Refuses to remove the extension while user-owned `Gateway` objects still exist in the shoot (to prevent silent traffic loss). The admission webhook surfaces this as a validation error on the `Shoot` update. Once the user has cleaned up their `Gateway`/`*Route` objects, the `ManagedResource` is deleted and the extension waits up to 5 minutes for managed objects to disappear. |

> | Delete (shoot deletion) | Shoot deletion bypasses the "Gateways still exist" guard — the entire shoot is going away anyway, so blocking would only leak the shoot. The extension's `Extension` resource is reconciled with `lifecycle.delete: BeforeKubeAPIServer`, so the `ManagedResource` (Envoy Gateway control plane, CRDs, the `envoy-gateway` `GatewayClass`, EnvoyProxy/HTTPRoute/Gateway instances) is torn down before the shoot's API server is removed. The cloud-provider `LoadBalancer` Services that were created for each `Gateway` are deleted as part of the shoot's normal `Service` cleanup, freeing the underlying load balancers. No manual cleanup of `Gateway` objects is required from the user. |

Reviewer 2: How is this case distinguished from the case before, i.e. how does the extension figure out if the shoot is deleted? Does it always check the `deletionTimestamp`?

| Heartbeat | Extension controller participates in the Gardener heartbeat protocol and reports liveness. |
| Metrics | Prometheus metrics are exposed on port 8080 under `/metrics`; Gardener's monitoring stack can scrape them via `ServiceMonitor`. The Envoy data-plane and Envoy Gateway control-plane also expose Prometheus metrics that are scraped via separate `ServiceMonitor` objects. |

> | VPA/HPA | Optional VPA and HPA manifests for the control-plane and data-plane pods are provided in the Helm chart (VPA enabled by default, HPA opt-in). |

Reviewer 2: Will VPA/HPA also be included for the envoy gateway components in the shoot cluster, i.e. will there be autoscaling available?

| C3 | **Architectural correctness**: control- vs data-plane separation, per-namespace Gateway isolation | Multi-team shoots require strong tenant isolation |
| C4 | **Status reporting timeliness** | Slow status updates break GitOps workflows |
| C5 | **Scale**: number of Gateways, Routes, backends supported on a single control plane | A shoot can host hundreds of routes |
| C6 | **Project governance**: vendor neutrality, CNCF status, multi-vendor maintainers | Avoid vendor lock-in |
| C7 | **Operational footprint**: control-plane components, dependency on a service mesh, CRD count | Operability inside a Gardener shoot |
| C8 | **NGINX/Ingress migration story**: support for users coming from `Ingress` | Backwards compatibility with GEP-57 Traefik ingress users |
| C9 | **L4 routing** (`TCPRoute`, `TLSRoute`, `UDPRoute`) | Gardener users running databases, gRPC, custom protocols |
| C10 | **Already in use within Gardener** | Reuse operational knowledge |

### Candidates Considered

The candidate list includes every implementation that ships a usable

> Gateway API standard-channel controller as of late 2025. Implementations

Reviewer 2: `every implementation` is a very strong statement that is likely to be disproven. Maybe `several common` is better.

shipping only the Ingress API are excluded.

> #### 1. Envoy Gateway

Reviewer 2: Let's add links to all the candidates.

Selected. See [Why Envoy Gateway Over Traefik Gateway API](#why-envoy-gateway-over-traefik-gateway-api).

* **Throughput** (gateway-api-bench, 512 conns): ~325k qps
* **Architecture**: Clean control-plane (Go) / data-plane (Envoy) split.
  Per-namespace Gateway isolation respected.
* **Conformance**: Full standard channel, partial experimental channel.
  No correctness issues reported in the upstream benchmark beyond a memory
  leak under churn.
* **Governance**: CNCF, hosted under the Envoy organisation. Multi-vendor
  maintainers (Tetrate, VMware, AIS, Bloomberg, others).
* **L4**: `TCPRoute`, `TLSRoute`, `UDPRoute` all supported.
* **Dependencies**: None beyond Kubernetes and the Gateway API CRDs.

#### 2. Traefik Gateway API

Runner-up. Rejected primarily on architectural correctness in multi-tenant
scenarios, on slow status reconciliation, and on benchmarked failures with
large route volumes.

* **Throughput** (gateway-api-bench): ~217k qps — solid mid-pack.
* **Architecture concern**: The upstream benchmark notes that Traefik
  *"consolidates all Gateways across namespaces unsafely"*, meaning a
  single Traefik process serves Gateways from multiple namespaces. In a

>   multi-tenant Gardener shoot, this is a tenant-isolation violation.

Reviewer 2: Maybe we should add that this is exactly how most `Ingress` controllers operate. Hence, it might be a good approach nevertheless depending on the scenario.

* **Status reporting**: ~180 seconds latency to reflect status on large route

>   sets — incompatible with GitOps tooling that polls for `Ready` conditions.

Reviewer 2: Why is this `incompatible with GitOps tooling`? Polling should eventually get the feedback.

      - [4. Kgateway (formerly Gloo Gateway, k8sgateway)](#4-kgateway-formerly-gloo-gateway-k8sgateway)
      - [5. Cilium Gateway API](#5-cilium-gateway-api)
      - [6. Kong Gateway](#6-kong-gateway)
      - [7. NGINX Gateway Fabric](#7-nginx-gateway-fabric)
    - [Final Decision Matrix](#final-decision-matrix)
    - [Why Envoy Gateway Over Traefik Gateway API](#why-envoy-gateway-over-traefik-gateway-api)
  - [Future Enhancements](#future-enhancements)
  - [Drawbacks](#drawbacks)
  - [Alternatives](#alternatives)
    - [1. Do Nothing — Rely on `shoot-traefik` Plus User-Installed Gateway API](#1-do-nothing--rely-on-shoot-traefik-plus-user-installed-gateway-api)
    - [2. Pick Traefik Gateway API for Operational Reuse](#2-pick-traefik-gateway-api-for-operational-reuse)
    - [3. Pick Istio for Best Conformance](#3-pick-istio-for-best-conformance)
    - [4. Pick Kgateway for Best Performance](#4-pick-kgateway-for-best-performance)
    - [5. Cilium Gateway API for CNI Co-location](#5-cilium-gateway-api-for-cni-co-location)
    - [6. NGINX Gateway Fabric for "Familiar" NGINX Branding](#6-nginx-gateway-fabric-for-familiar-nginx-branding)
    - [7. Kong Gateway](#7-kong-gateway)

>     - [8. Use Gateway API CRDs Without a Bundled Implementation](#8-use-gateway-api-crds-without-a-bundled-implementation)

Reviewer 4: This is no longer part of the template.

Reviewer: Please, remove the table of content

  extension and this extension can be installed in the same shoot. They
  do not conflict at the IngressClass / GatewayClass level. See
  [Coexistence with `shoot-traefik`](#coexistence-with-shoot-traefik).

* **Scope is currently restricted to `purpose: evaluation` shoots.** This
  constraint is enforced by the admission webhook and exists to allow the
  extension to mature in low-risk environments before being opened up to
  development and production shoots.

### Risks and Mitigations

| Risk | Likelihood | Impact | Mitigation |
|------|-----------|--------|------------|
| Envoy Gateway memory leak under route/config churn (reported in [gateway-api-bench](https://github.com/howardjohn/gateway-api-bench)) | Medium | Medium | The control-plane pod is configured with a memory limit and a VPA `Recreate` updatePolicy so leaked memory is reclaimed by periodic restart. The issue is actively tracked upstream; the extension pins to a release where mitigations have landed. |
| Initial-request errors during bootstrap (also reported upstream) | Medium | Low | Readiness gates on the data-plane Envoy pods; documentation advises users to use a startup-probe-based health check from their LB. |

> | Gateway API CRD conflicts if user pre-installed them | Low | High | CRDs are part of the shoot `ManagedResource` and applied via server-side apply; existing CRDs are updated idempotently without field-ownership conflicts. The extension's CRD reconciliation is opt-out via `spec.manageCRDs: false` for users who manage CRDs themselves. |

Reviewer 4: Furthermore, this is still opt-in, right, i.e., users are aware of this and can manage their own Gateway API deployments accordingly.

metadata:
  name: gardener-extension-shoot-envoy-gateway
helm:
  rawChart: <base64-encoded Helm chart>
```

```yaml
# ControllerRegistration
apiVersion: core.gardener.cloud/v1beta1
kind: ControllerRegistration
metadata:
  name: shoot-envoy-gateway
spec:
  resources:
    - kind: Extension
      type: shoot-envoy-gateway
      globallyEnabled: false   # opt-in per shoot
      lifecycle:
        reconcile: AfterKubeAPIServer
        delete:   BeforeKubeAPIServer
  deployment:
    deploymentRefs:
      - name: gardener-extension-shoot-envoy-gateway
```

The extension controller is deployed per seed and watches `Extension` objects
of type `shoot-envoy-gateway`.

### API

The extension introduces a new API group `envoy-gateway.extensions.gardener.cloud`
with a single versioned kind `EnvoyGatewayConfig`:

```yaml
apiVersion: envoy-gateway.extensions.gardener.cloud/v1alpha1
kind: EnvoyGatewayConfig
spec:
  # Number of Envoy Gateway control-plane replicas (default: 2)
  controlPlaneReplicas: 2

  # Number of Envoy data-plane (proxy) replicas per Gateway (default: 2)
  dataPlaneReplicas: 2

  # Log level for both control-plane and data-plane: debug | info | warn | error (default: info)

>   logLevel: info

Reviewer 4: Usually, we don't expose such fields and rather let auto-scaling handle this. Can we follow the same approach?

  # Manage (install and update) the Gateway API CRDs. Set to false if CRDs
  # are owned externally. (default: true)

>   manageCRDs: true

Reviewer 4: See above, why would we need this? If Gardener shall deploy a Gateway API extension, I guess it's reasonable to force that Gardener also deploys the corresponding CRDs?

  # Install the experimental-channel Gateway API CRDs (TCPRoute, TLSRoute, UDPRoute,
  # BackendTLSPolicy). (default: false)
  experimentalFeatures: false

  # Optional pinned EnvoyProxy template applied to every Gateway via the
  # `gateway.envoyproxy.io/v1alpha1.EnvoyProxy` reference.
  envoyProxyDefaults:
    resources:
      requests:
        cpu: 100m

>         memory: 256Mi

Reviewer 4: same here?

  template fields (resources, access logs, tracing, rate-limit backend)
  through `EnvoyGatewayConfig` so shoot owners can tune common knobs without
  hand-writing `EnvoyProxy` objects.

* **Migration tooling.** A helper command (`gardenctl ingress-to-gateway`)
  that converts NGINX-style `Ingress` and Traefik `IngressRoute` objects
  into equivalent `Gateway` + `HTTPRoute` resources for users migrating off
  GEP-57.

* **Re-evaluation of Kgateway.** Once Kgateway graduates from CNCF sandbox,
  re-run the evaluation. Its raw throughput (~400k qps) is the highest of
  the field, and it is built on the same Envoy data plane. Note that any
  potential adoption would *not* be a drop-in switch: although both
  implementations sit on Envoy, their CRDs (`EnvoyProxy`,
  `BackendTrafficPolicy`, `SecurityPolicy`, …) are not API-compatible, and
  shoot users who rely on Envoy-Gateway-specific extensions would need a
  documented migration path. A switch is therefore conditional on (a)
  Kgateway graduating to CNCF incubation and (b) a clear plan for migrating
  existing user-facing configuration. If those preconditions are not met, the
  extension stays on Envoy Gateway.

## Drawbacks

> * **Two ingress extensions to maintain.** Gardener will end up with both

Reviewer 4: Shouldn't the long-term plan be to deprecate the `shoot-traefik` extension and support? We have developed it as stop-gap solution, but Gateway API is _the_ way to go, right?

Author: Honestly I would offer both. The `shoot-traefik` extension as an Ingress replacement is the simplest solution to use and covers most HTTP use cases — as long as it's maintained, there's no reason to drop it or not to work on production qualities for the extension. It's the path of least resistance for the 80% of workloads that just need host/path → service routing, and the ecosystem around it. Gateway API should be available as an option for new workloads that need more than what Ingress can express. @ScheererJ what do you think?

   install one independently; because the mesh consumes the standard-channel
   route CRDs this extension already manages, the two compose cleanly. Adding
   first-class GAMMA support to this extension is noted under
   [Future Enhancements](#future-enhancements).
4. This GEP does **not** introduce multiple competing `GatewayClass`
   objects. The extension installs a single `GatewayClass` named
   `gardener-envoy-gateway` bound to the Envoy Gateway controller. Users may
   install additional `GatewayClass` objects pointing at other implementations
   independently of this extension.
5. This GEP does **not** cover network policy, mTLS automation, or
   advanced traffic policies (rate limiting, JWT auth, WAF), nor automated
   integration with other Gardener extensions such as `shoot-dns-service`
   (DNS records) or `shoot-cert-service` (TLS certificates) for
   Gateway-exposed workloads. Gateway API resources can reference
   externally-managed DNS names and TLS secrets today, but wiring those
   extensions to react to `Gateway`/`HTTPRoute` objects is not on the initial
   roadmap. These can be layered on top via the chosen implementation's policy

>    CRDs but are not exercised by the extension itself in the initial release.

Reviewer: > These can be layered on top via the chosen implementation's policy CRDs but are not exercised by the extension itself in the initial release.

I do not understand this sentence, maybe it can be removed?

Author: Advanced traffic and security features like rate limiting, JWT auth, WAF, DNS/cert automation, etc can be added by creating Envoy Gateway's own policy CRDs for example SecurityPolicy, BackendTrafficPolicy, etc. on top of the standard Gateway/HTTPRoute resources. However, the extension itself won't create or manage those policies in the initial release. The CRDs are available for users to author, but the extension stays hands-off. If this is not clear, I can remove the sentence. :)

Reviewer: I get it now, maybe it would better to split this paragraph in two, one for the integration with dns and cert extensions, and one for the advanced features of envoy gateway. I was really thinking this sentence was referring to the dns and cert extensions and in this context I did not understand it, but now with the explanation it is about the advanced envoy features, it makes sense. So either rephrase the paragraph to make it more clear or split it into two.

Author: I've split it now into two paragraphs as suggested. Hope this makes it now clearer. :)

  experimental CRDs (`TCPRoute`, `TLSRoute`, `UDPRoute`, `BackendTLSPolicy`)
  are not installed by default. They can be enabled by setting
  `channel: experimental` in the extension's provider config
  (`spec.extensions[].providerConfig` on the Shoot, see [API](#api)).
  Operators should be aware that experimental APIs may change in
  backwards-incompatible ways between releases. To make this risk visible at
  the point of use, the admission webhook returns a non-fatal **warning** on
  every `Shoot` create/update that opts into `channel: experimental`;
  `kubectl` surfaces it inline without blocking the operation.

* **Envoy Gateway CRDs (e.g. `EnvoyProxy`, `BackendTrafficPolicy`,
  `ClientTrafficPolicy`, `SecurityPolicy`) are also installed.** These are
  namespaced and required for advanced Envoy-specific configuration. Users
  who only consume the standard Gateway API surface can ignore them.

* **Where the workload runs in the shoot.** The Envoy Gateway control plane
  (Deployment, Service, RBAC, PDB, VPA/HPA) is deployed into the
  `kube-system` namespace in the shoot. The per-`Gateway` Envoy
  data-plane proxy Deployments and their `LoadBalancer` Services are created

>   by the Envoy Gateway control plane in the namespace of the corresponding

Reviewer: Is it possible to run the dataplane proxies in the kube-system namespace? Usually kube-system is reserved for system components and service providers (Gardener), having gardener managed/originating workload outside the kube-system namespace may be unexpected.

Author: Yes it's possible. I've adapted the GEP to run them in the `kube-system` namespace. :)

# GEP-68: Gateway API Extension for Gardener Shoot Clusters

> ## Table of Contents

Reviewer 5: The table of contents is no longer part of the GEP template. Please remove it in favor of GitHub's built-in document outline.

* **Optional role separation**: Some Gardener shoots are shared by multiple
  teams where platform admins and application developers have distinct
  responsibilities. Gateway API's persona-based resource model
  (`GatewayClass` → `Gateway` → `*Route`) enables a separation of concerns
  that `Ingress` does not offer natively. For single-team shoots or simpler
  setups where this split is unnecessary, `Ingress` remains a
  straightforward and perfectly valid choice.

* **Vendor neutrality and conformance**: Gateway API has a published
  conformance suite, and workloads that stick to the standard-channel
  resources (`HTTPRoute`, `GRPCRoute`, etc.) are portable across conformant
  implementations in principle. In practice, however, most non-trivial
  deployments will rely on implementation-specific extensions — Envoy
  Gateway's `EnvoyPatchPolicy`/`EnvoyExtensionPolicy`, Istio's
  `EnvoyFilter`, etc. — which limits real-world portability. The
  conformance suite nonetheless provides a useful baseline: the standard
  routing semantics are well-defined, and simple-to-moderate use cases
  transfer across implementations without modification.

### Goals

> 1. Introduce `gardener-extension-shoot-envoy-gateway` as an extension in the

Reviewer 4: ```suggestion
1. Introduce `gardener-extension-envoy-gateway` as an extension in the
```

I'm not a fan of this `shoot` infix and would simply drop it.

   [Gardener GitHub organisation](https://github.com/gardener).
2. Ship **Envoy Gateway** as the bundled Gateway API implementation. Because
   the upstream Envoy Gateway Helm chart does not ship a `GatewayClass`, the
   extension installs one named `envoy-gateway` bound to the Envoy Gateway
   controller (`gateway.envoyproxy.io/gatewayclass-controller`). Shoot users
   reference it from their `Gateway` resources via
   `spec.gatewayClassName: envoy-gateway`.
3. Install (or reconcile) the Gateway API standard channel CRDs
   (`gateway.networking.k8s.io/v1`) in the shoot cluster.
4. Integrate with Gardener's standard resource-management, observability,
   and lifecycle mechanisms (`ManagedResource`, heartbeat, metrics, VPA/HPA).
5. Provide an opinionated baseline configuration that works out of the box
   on every cloud provider Gardener supports, while allowing users to
   override behaviour through the extension's `providerConfig`.
6. Enable safe, incremental rollout: restrict the extension initially to
   shoots with `purpose: evaluation`, mirroring the rollout strategy of

>    GEP-57.

Reviewer 5: Should we make this restriction configurable by operators on the extension?

* **Architectural cleanliness**: Proper control plane / data-plane separation,
  per-namespace Gateway isolation that respects the Gateway API spec.
* **Conformance**: Full support for the Gateway API standard channel; no
  reported correctness issues in the upstream benchmark beyond a known
  memory leak under churn (see [Risks and Mitigations](#risks-and-mitigations)).
* **Foundation**: Built on [Envoy](https://www.envoyproxy.io/), the same
  data plane used by [Istio](https://istio.io/),
  [Kgateway](https://kgateway.dev/), and most cloud-native gateway and
  service-mesh products — a widely adopted and well-proven data path in
  the Kubernetes ecosystem.
* **CNCF stewardship**: Hosted under the [Envoy organisation](https://www.envoyproxy.io/community)
  in [CNCF](https://www.cncf.io/), with multi-vendor maintainers — no
  single-vendor lock-in.

Traefik Gateway API was a strong runner-up but rejected on architectural
correctness in multi-tenant scenarios, slow status reconciliation, and
benchmarked failures with large route volumes — see
[Why Envoy Gateway Over Traefik Gateway API](#why-envoy-gateway-over-traefik-gateway-api).

### Notes/Constraints/Caveats

> * **Gateway API CRDs are installed cluster-wide.** The extension installs the

Reviewer 5: This means that if shoot owners want to use another Gateway API implementation in parallel, they will need to disable management of the CRDs and accept the version deployed by this extension, correct?

Author: Yes

  Operators should be aware that experimental APIs may change in
  backwards-incompatible ways. To make this risk visible at the point of
  use, the admission webhook returns a non-fatal **warning** on every
  `Shoot` create/update that opts into `experimentalFeatures: true` (see
  [Admission Webhook](#admission-webhook)); `kubectl` surfaces it inline
  without blocking the operation.

* **Envoy Gateway CRDs (e.g. `EnvoyProxy`, `BackendTrafficPolicy`,
  `ClientTrafficPolicy`, `SecurityPolicy`) are also installed.** These are
  namespaced and required for advanced Envoy-specific configuration. Users
  who only consume the standard Gateway API surface can ignore them.

* **`GatewayClass` is `envoy-gateway`.** The extension installs a single
  `GatewayClass` named `envoy-gateway` bound to controller
  `gateway.envoyproxy.io/gatewayclass-controller` (see [Goals](#goals)). Users
  reference it from their `Gateway` objects via
  `spec.gatewayClassName: envoy-gateway`.

* **Shoot resources are authored as Go types, not Helm.** Following the
  Gardener
  [component checklist](https://github.com/gardener/gardener/blob/master/docs/development/component-checklist.md),

>   the workload deployed into the shoot (control plane `Deployment`, `Service`,

Reviewer: In which namespace(s) are they deployed?

Author: in kube-system ns

| Initial-request errors during bootstrap (also reported upstream) | Medium | Low | Readiness gates on the data-plane Envoy pods; documentation advises users to use a startup-probe-based health check from their LB. |
| Gateway API CRD conflicts if user pre-installed them | Low | High | CRDs are part of the shoot `ManagedResource` and applied via server-side apply; existing CRDs are updated idempotently without field-ownership conflicts. The extension's CRD reconciliation is opt-out via `spec.manageCRDs: false` for users who manage CRDs themselves. |
| Limited annotation/feature parity with NGINX or Traefik Ingress | High | Medium | Documented migration guide: most NGINX `Ingress` annotations have a direct `HTTPRoute` filter or Envoy `BackendTrafficPolicy` equivalent. Users requiring features not yet expressible in standard Gateway API are advised to remain on the Traefik extension until the experimental channel covers their case. |

> | Operators end up with three concurrent ingress paths in one shoot (legacy NGINX, Traefik, Gateway API) | Medium | Medium | Documentation strongly recommends a single ingress path per shoot in production. The evaluation-purpose scope of both extensions limits the blast radius during the rollout phase. |

Reviewer: Please, add a reference to the documentation recommending single ingress path per shoot.

| Ecosystem churn: Gateway API evolves rapidly (v1.5 as of early 2026, v1.6 in progress) | High | Low | The extension declares conformance against a specific Gateway API release in its release notes and bumps deliberately, not automatically. When a shoot opts into the experimental channel, the admission webhook additionally emits a non-fatal warning on create/update flagging the backwards-incompatibility risk (see [Admission Webhook](#admission-webhook)). |

## Design Details

> ### Extension Registration

Reviewer 5: This is a personal preference, but this section doesn't provide any new information beyond the extension lifecycle. So, it could be condensed to a single sentence.

      reconcile: AfterKubeAPIServer
    type: shoot-envoy-gateway
    workerlessSupported: false
```

or as `ControllerDeployment` and `ControllerRegistration`:

```yaml
# ControllerDeployment
apiVersion: core.gardener.cloud/v1beta1
kind: ControllerDeployment
metadata:
  name: gardener-extension-shoot-envoy-gateway
helm:
  rawChart: <base64-encoded Helm chart>
```

```yaml
# ControllerRegistration
apiVersion: core.gardener.cloud/v1beta1
kind: ControllerRegistration
metadata:
  name: shoot-envoy-gateway
spec:
  resources:
    - kind: Extension
      type: shoot-envoy-gateway
      globallyEnabled: false   # opt-in per shoot
      lifecycle:
        reconcile: AfterKubeAPIServer
        delete:   BeforeKubeAPIServer
        migrate: AfterKubeAPIServer
  deployment:
    deploymentRefs:
      - name: gardener-extension-shoot-envoy-gateway
```

The extension controller is deployed per seed and watches `Extension` objects
of type `shoot-envoy-gateway`.

### API

The extension introduces a new API group `envoy-gateway.extensions.gardener.cloud`
with a single versioned kind `EnvoyGatewayConfig`:

```yaml
apiVersion: envoy-gateway.extensions.gardener.cloud/v1alpha1
kind: EnvoyGatewayConfig

> spec:

Reviewer 5: Usually, we don't have a `spec` section inside `providerConfig` APIs (as there is no `status` section either)

Author: removed it, thanks for the hint.

  # Number of Envoy Gateway control plane replicas (default: 2)
  controlPlaneReplicas: 2

  # Number of Envoy data-plane (proxy) replicas per Gateway (default: 2)

>   dataPlaneReplicas: 2

Reviewer: If you are already envisioning options to configure other control/data plane features, it should be nice to group them in the API per "plane" (I would say you can have the log level setting per "plane"). 
```suggestion
  # Number of Envoy Gateway control plane replicas (default: 2)
  controlPlane:
    replicas: 2
    logLevel: ...

  # Number of Envoy data-plane (proxy) replicas per Gateway (default: 2)
  dataPlane:
    replicas: 2
    logLevel: ...
```

  # Log level for both control plane and data-plane: debug | info | warn | error (default: info)
  logLevel: info

  # Manage (install and update) the Gateway API CRDs. Set to false if CRDs
  # are owned externally. (default: true)
  manageCRDs: true

  # Install the experimental-channel Gateway API CRDs (TCPRoute, TLSRoute, UDPRoute,
  # BackendTLSPolicy). (default: false)

>   experimentalFeatures: false

Reviewer 5: Did you consider a `channel=standard|experimental` field instead? "Experimental features" could also mean something else than just selecting the channel.

Author: renamed it

        experimentalFeatures: false
  # ... remaining shoot spec
```

To enable the experimental Gateway API CRDs, set
`spec.experimentalFeatures: true` in the provider config above.

### Provider Modes

The extension does not expose multiple back-end providers. Unlike the
`shoot-traefik` extension, where `KubernetesIngress` vs `KubernetesIngressNGINX`
modes serve a clear migration purpose, Gateway API is a single, well-defined
spec — there is no analogous compatibility shim to model. The implementation
behind the `envoy-gateway` `GatewayClass` is always Envoy Gateway in this release.

A user wishing to run a different implementation (for example Istio for mesh
integration) can do so independently of this extension by installing their
own `GatewayClass`. Such side-by-side installations are out of scope for the
extension's own validation but are not blocked.

### Admission Webhook

A `ValidatingWebhookConfiguration` is registered in the garden cluster (the
virtual garden), where `Shoot` resources live, at path
`/webhooks/validate-shoot-envoy-gateway`. It validates incoming `Shoot`
objects:

1. **Purpose restriction** (temporary): The extension may only be enabled

>    (`disabled: false`) on Shoots whose `spec.purpose` is `evaluation`. Adding

Reviewer 4: What does this `disabled: false` mean here?

Reviewer: I guess the LLM just was too detailed and outlined that `shoot.spec.extensions[].disabled=false` means that the extension is enabled.

Reviewer 4: Aha, thanks, seems somewhat reasonable. Still, confusing. I'd remove it.

   accepted enum values.

Beyond hard rejections, the webhook also emits **non-fatal warnings** via the
`AdmissionResponse.warnings` field (surfaced inline by `kubectl` and other
clients on create/update, without blocking the request):

* **Experimental channel opt-in**: When `spec.experimentalFeatures: true` is
  set, the webhook warns that the experimental-channel Gateway API CRDs
  (`TCPRoute`, `TLSRoute`, `UDPRoute`, `BackendTLSPolicy`) may change in
  backwards-incompatible ways between releases, so that shoot owners are
  reminded of the stability trade-off every time they apply the manifest —
  not just when they first read the docs.
* **Coexisting ingress paths**: When both `shoot-traefik` and
  `shoot-envoy-gateway` are enabled on the same shoot, the webhook warns
  about the doubled load-balancer cost and recommends a single ingress path
  per shoot in production (see
  [Coexistence with `shoot-traefik`](#coexistence-with-shoot-traefik)).

Warnings are advisory only; they never fail the admission request. This keeps
the experimental-features risk visible at the point of use while leaving the
choice with the shoot owner.

> ### Lifecycle Management

Reviewer 5: TBH, I didn't read this section because it goes too much into implementation details. Should we remove it from this design document?

Reviewer 4: Yes, please.

Author: removed it

### Coexistence with `shoot-traefik`

A shoot may have both `shoot-traefik` and `shoot-envoy-gateway` enabled. The
two extensions reconcile disjoint resources:

| Resource | Owner |
|----------|-------|
| `IngressClass: nginx` / `IngressClass: traefik` | `shoot-traefik` |
| `GatewayClass: envoy-gateway` | `shoot-envoy-gateway` (registered by Envoy Gateway itself) |
| Traefik Deployment, IngressRoute CRDs | `shoot-traefik` |
| Envoy Gateway Deployment, Gateway API CRDs, EnvoyProxy CRDs | `shoot-envoy-gateway` |

There is no IP/port conflict, because the two paths never share a `Service`:
`shoot-traefik` provisions its own `LoadBalancer` Service for the Traefik
proxy, while under `shoot-envoy-gateway` a `LoadBalancer` Service is created
per user `Gateway` by the Envoy Gateway control plane (see
[Deployment Topology](#deployment-topology-seed-vs-shoot)) — the extension
itself provisions none. The cost consideration is therefore conditional: a
shoot that actually runs `Gateway` objects alongside Traefik ingress pays for
both sets of load balancers.

> ### Scope Restriction: Evaluation Shoots Only

Reviewer 5: I feel like this (similar to some other statements) is repeated too often in this document 😄

Traefik in shoots — does not carry over to the Gateway API surface. The
GEP-57 extension exists to provide a smooth migration off the retired
NGINX Ingress controller, and its value lies in NGINX annotation
translation, which is wholly an Ingress-API feature. Gateway API users do
not benefit from that translation. Therefore the operational-reuse argument
applies only to the binary, not to the user-facing API — and even there,
the two extensions can coexist (see
[Coexistence with `shoot-traefik`](#coexistence-with-shoot-traefik)).

For these reasons the extension ships Envoy Gateway. Operators who prefer
Traefik for the Gateway API surface can install it independently in their
shoots and register a separate `GatewayClass`; this GEP does not block
them.

## Future Enhancements

* **Promote out of `purpose: evaluation`.** Once the extension has soaked in
  evaluation shoots and the upstream Envoy Gateway memory-leak issue is
  resolved, the admission-webhook restriction will be removed.

* **Feature gates** for opt-in experimental Gateway API features
  (`TCPRoute`, `BackendTLSPolicy`, mesh GAMMA bindings) without requiring a

>   new extension release.

Reviewer 4: I don't understand this point. I thought this happens when `.spec.experimentalFeatures=true`? Why would it require a new extension release?

* **Implementation-version decoupling.** In the initial release, the Envoy
  Gateway version is pinned in the extension binary. A future iteration
  will allow operators to manage a catalog of Envoy Gateway versions with
  lifecycle classifications (aligned with
  [GEP-32](../0032-version-classification-lifecycle/README.md)) and let

>   shoot owners pin a specific version.

Reviewer 4: You should probably mention that this is a separate, dedicated topic for another GEP that you plan to piggy-back on.

* **GatewayClass parameters.** Expose a curated set of `EnvoyProxy`
  template fields (resources, access logs, tracing, rate-limit backend, and
  common load-balancer knobs such as internal/private exposure) through
  `EnvoyGatewayConfig` so shoot owners can tune common knobs without
  hand-writing `EnvoyProxy` objects or per-`Gateway`

>   `spec.infrastructure.annotations`.

Reviewer 4: Can you explain this and/or make an example what you have in mind here?

Author: Rephrased it

* **First-class GAMMA (service mesh) support.** A future iteration could
  integrate a mesh data plane (for example Envoy Gateway's mesh mode or a
  co-located ambient mesh) so shoot owners can attach `HTTPRoute`/`GRPCRoute`
  objects directly to `Service` resources without operators having to install
  and run a separate mesh. Why this is out of scope today is covered in
  [Non-Goals](#non-goals).

> * **Migration tooling.** A helper command (`gardenctl ingress-to-gateway`)

Reviewer 5: I don't think `gardenctl` is a good fit for such migration tooling. I prefer a dedicated binary in this extension, a dedicated repository, or even just a simple GitHub gist.

  documented migration path. A switch is therefore conditional on (a)
  Kgateway graduating to CNCF incubation and (b) a clear plan for migrating
  existing user-facing configuration. If those preconditions are not met, the
  extension stays on Envoy Gateway.

## Drawbacks

* **Two ingress extensions to maintain.** Gardener will end up with both
  `shoot-traefik` (Ingress API) and `shoot-envoy-gateway` (Gateway API).
  This doubles the maintenance surface and creates a "which one do I
  use?" question for new users. Documentation will need to address this
  clearly, and there is a real risk of operator confusion in the
  transition window.

* **Envoy Gateway is younger than Traefik.** Envoy Gateway 1.0 shipped in
  2024. Although Envoy itself is mature, the Gateway-API-specific control
  plane has fewer production-years behind it than Traefik. The known
  memory-leak issue is one symptom of this.

* **CRD proliferation.** The extension installs Gateway API CRDs (5
  standard-channel + up to 4 experimental) and Envoy Gateway CRDs (5+).
  Even a user who only deploys a single `HTTPRoute` will see >10 CRDs
  appear in their shoot. This is unavoidable for any Gateway API

>   implementation and is shared with all Gateway API extensions.

Reviewer 4: Why do you mention this as drawback? That's unavoidable and doesn't come with any negative consequences, or? Remove it?

* **Coexistence cost.** Running both `shoot-traefik` and
  `shoot-envoy-gateway` in the same shoot doubles the load-balancer cost.

>   Operators need to communicate this to shoot owners.

Reviewer 4: Why operators? I thought you print a warning in your admission controller?

* **No annotation-compat shim for Gateway API.** Unlike `shoot-traefik`'s
  `KubernetesIngressNGINX` mode, this extension does not translate `Ingress`
  resources or NGINX annotations into `HTTPRoute`. Users migrating from
  `Ingress` will need to rewrite their routing manifests. A future

>   migration tool is mentioned in [Future Enhancements](#future-enhancements).

Reviewer 4: Isn't this a duplicate of your migration point from above?

Author: removed it.

### 5. Cilium Gateway API for CNI Co-location

The argument: shoots using Cilium as CNI could collapse CNI and Gateway into a
single component. **Rejected** on the benchmarked performance and
silent-configuration-failure behaviour detailed under candidate
[Cilium Gateway API](#5-cilium-gateway-api).

### 6. NGINX Gateway Fabric for "Familiar" NGINX Branding

The argument: continuity for users coming from Ingress NGINX. **Rejected** —
it is a separate codebase from the retired `ingress-nginx` (no real
continuity), and the benchmark flags both low throughput and crashes during
route changes. See candidate [NGINX Gateway Fabric](#7-nginx-gateway-fabric).

### 7. Kong Gateway

**Rejected** on the same multi-tenant namespace-consolidation issue as Traefik
plus incorrect route-count reporting. See candidate
[Kong Gateway](#6-kong-gateway).

### 8. Use Gateway API CRDs Without a Bundled Implementation

The argument: install only the Gateway API CRDs and a `GatewayClass`
registry; let users pick their implementation. **Rejected** because it
provides no working ingress out of the box, contradicting the goal of a
turnkey extension. This is functionally equivalent to "do nothing" for

> most users.

Reviewer 5: Although I agree with your point that this wouldn't be a turnkey solution anymore, operators/service providers could still benefit from this mode.
In the STACKIT Kubernetes Engine, we (will) have multiple controllers/extensions working with Gateway API. Hence, we will need a central/common extension that manages the Gateway API CRDs that is independent of the feature implementations. 

Would it be an option to add another mode for this extension to only manage Gateway API CRDs?

Reviewer 2: From my point of view, it would complicate things unnecessarily if the extensions tries to cater to a fully gateway api implementation, i.e. CRDs + envoy-gateway, and a case of only installing the CRDs. The former case is clearly coupled. Hence, it gets complicated when envoy-gateway does not adopt a new gateway api release right away, but someone using the second scenario wants it.
If there is a need to "just" install the gateway api CRDs into a cluster it should be a separate extension. Feel free to provide one if that is anyway on your roadmap.

Reviewer: For CRDs deployment only probably [static manifests](https://github.com/gardener/gardener/blob/master/docs/extensions/static-manifests.md) can serve the purpose.

| VPA/HPA | Optional VPA and HPA manifests are provided for both the Envoy Gateway control plane and the Envoy data-plane pods in the shoot cluster (VPA enabled by default, HPA opt-in). The extension controller on the seed has its own VPA configuration (see [Extension Registration](#extension-registration)). |

The admission webhook (registered in the garden cluster where `Shoot`
resources live) validates the `EnvoyGatewayConfig` `providerConfig` (strict
decoding, field constraints such as replica counts ≥ 1 and a valid
`logLevel`/`channel`) and enforces the temporary `purpose: evaluation` scope.
It additionally emits non-fatal **warnings** — surfaced inline by `kubectl`
without blocking the request — when `channel: experimental` is selected
(flagging the backwards-incompatibility risk) and when both `shoot-traefik`
and `envoy-gateway` are enabled on the same shoot (flagging the doubled
load-balancer cost).

### Coexistence with `shoot-traefik`

A shoot may have both `shoot-traefik` and `envoy-gateway` enabled. The
two extensions reconcile disjoint resources:

| Resource | Owner |
|----------|-------|
| `IngressClass: nginx` / `IngressClass: traefik` | `shoot-traefik` |

> | `GatewayClass: envoy-gateway` | `envoy-gateway` (registered by Envoy Gateway itself) |

Reviewer 5: Should we choose a non-default `GatewayClass` name for the one managed by the extension?

Author: I've renamed it to `gardener-gateway`
